riscv/src/compiler.rs

```diff
@@ -635,7 +635,78 @@ fn preamble(degree: u64, coprocessors: &CoProcessors, with_bootloader: bool) ->
+    //   have a read as the first operation on a new address.
+    // - The `(1 - m_change) * LAST = 0` constraint is replaced with
+    //   `LAST * (1 - m_change) * (m_addr + 1) = 0`. This allows for a valid assignment in the case
+    //   where there is no memory operation in the entire chunk: The address can be set to -1 (which
+    //   cannot be represented in 32 bits, hence there is can't be an actual memory operation
+    //   associated with it).
+    let bootloader_specific_parts = if with_bootloader {
+        r#"
+    // If the operation is a write operation.
+    col witness m_is_write;
+    col witness m_is_bootloader_write;
+    col witness m_is_read;
+
+    // All operation flags are boolean and either all 0 or exactly 1 is set.
+    m_is_write * (1 - m_is_write) = 0;
+    m_is_read * (1 - m_is_read) = 0;
+    m_is_bootloader_write * (1 - m_is_bootloader_write) = 0;
+    m_is_read * m_is_write = 0;
+    m_is_read * m_is_bootloader_write = 0;
+    m_is_bootloader_write * m_is_write = 0;
+
+    // The first operation of a new address has to be a bootloader write
+    m_change * (1 - m_is_bootloader_write') = 0;
+
+    // m_change has to be 1 in the last row, so that the above constraint is triggered.
+    // An exception to this when the last address is -1, which is only possible if there is
+    // no memory operation in the entire chunk.
+    // This exception is necessary so that there can be valid assignment in this case.
+    pol m_change_or_no_memory_operations = (1 - m_change) * (m_addr + 1);
+    LAST * m_change_or_no_memory_operations = 0;
+
+    // If the next line is a read and we stay at the same address, then the
+    // value cannot change.
+    (1 - m_is_write' - m_is_bootloader_write') * (1 - m_change) * (m_value' - m_value) = 0;
+
+    /// Like mstore, but setting the m_is_bootloader_write flag.
+    instr mstore_bootloader Y, Z {
+        { X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000, STEP, Z } is m_is_bootloader_write { m_addr, m_step, m_value },
+        // Wrap the addr value
+        Y = (X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000) + wrap_bit * 2**32
+    }
+"#
+    } else {
+        r#"
+    // If the operation is a write operation.
+    col witness m_is_write;
+    col witness m_is_read;
+
+    // All operation flags are boolean and either all 0 or exactly 1 is set.
+    m_is_write * (1 - m_is_write) = 0;
+    m_is_read * (1 - m_is_read) = 0;
+    m_is_read * m_is_write = 0;
+
+    // m_change has to be 1 in the last row, so that a first read on row zero is constrained to return 0
+    (1 - m_change) * LAST = 0;
+
+    // If the next line is a read and we stay at the same address, then the
+    // value cannot change.
+    (1 - m_is_write') * (1 - m_change) * (m_value' - m_value) = 0;
+
+    // If the next line is a read and we have an address change,
+    // then the value is zero.
+    (1 - m_is_write') * m_change * m_value' = 0;
```

Comment:
This is moved code; if `with_bootloader` is `false`, the memory machine is the same as before.